src/main/java/org/spongepowered/api/util/blockray/RayTrace.java

```diff
@@ -0,0 +1,339 @@
+ * case where multiple predicates should be checked:</p>
+ *
+ * <ol>
+ *    <li>{@link #continueWhileLocation(Predicate)} -- called when the boundary
+ *    between two blocks is crossed, checking if the {@link ServerLocation} the
+ *    ray is crossing in to is valid, regardless of what the ray has otherwise
+ *    hit.</li>
+ *    <li>{@link #select(Predicate)} -- called when the {@link AABB} of
+ *    {@code T} is crossed by the ray trace, for {@link LocatableBlock}s this
+ *    is called directly after {@link #continueWhileLocation(Predicate)}. This
+ *    is the only predicate that can return a result, <strong>any</strong> .</li>
+ *    <li>{@link #continueWhileBlock(Predicate)} -- called if no selection was
+ *    returned (if appropriate), checks the {@link LocatableBlock} the ray is
+ *    entering. All predicates should return {@code true} if the ray should
+ *    continue to trace through the block.</li>
+ *    <li>{@link #continueWhileEntity(Predicate)} -- called if no selection was
+ *    returned (if appropriate), checks the {@link Entity} the ray has struck.
+ *    All predicates should return {@code true} if the ray should continue to
+ *    trace through the entity.</li>
+ * </ol>
+ *
+ * @param <T> The type of {@link Locatable} that this ray trace will attempt to
+ *            select
+ */
+public interface RayTrace<T extends Locatable> {
+
+    /**
+     * Creates a {@link RayTrace} that will attempt to select a
+     * {@link LocatableBlock}.
+     *
+     * @return The ray trace builder.
+     */
+    static RayTrace<LocatableBlock> block() {
+        return Sponge.getRegistry().getFactoryRegistry().provideFactory(Factory.class).blockRayTrace();
+    }
+
+    /**
+     * Creates a {@link RayTrace} that will attempt to select an {@link Entity}.
+     *
+     * @return The ray trace builder.
+     */
+    static RayTrace<Entity> entity() {
+        return Sponge.getRegistry().getFactoryRegistry().provideFactory(Factory.class).entityRayTrace();
+    }
+
+    /**
+     * Returns a {@link Predicate} for use in
+     * {@link #continueWhileLocation(Predicate)} that only allows a trace to
+     * continue if the {@link LocatableBlock} at the given position is air.
+     *
+     * @return The predicate
+     */
+    static Predicate<ServerLocation> onlyAir() {
+        return Sponge.getRegistry().getFactoryRegistry().provideFactory(Factory.class).onlyAir();
+    }
+
+    /**
+     * Returns a {@link Predicate} for use in {@link #select(Predicate)} that
+     * selects non-air blocks.
+     *
+     * @return The predicate
+     */
+    static Predicate<LocatableBlock> selectOnlyAir() {
+        return Sponge.getRegistry().getFactoryRegistry().provideFactory(Factory.class).selectNonAirBlock();
+    }
+
+    /**
+     * Sets the {@link ServerWorld} to perform the ray trace in.
+     *
+     * @param serverWorld The {@link ServerWorld}
+     * @return This, for chaining
+     */
+    RayTrace<T> world(ServerWorld serverWorld);
```

Comment:
Would there be a reason for ray tracing on the client thread at all?

Reply:
Quite possibly for if you want to do some client rendering stuff. It would probably require me to split the ray trace up into a `Client` and `Server` component as specifying a world is (probably) unneeded on the client, for example - it certainly doesn't have a notion of a world tied to a key at the very least.

Worth doing, do you think, even if I just lay the groundwork to enable writing a client based version later on down the road?